feat: add `esbuild` plugin

nedjulius opened the pull request:
Addresses issue #83 

Adds a _StyleX_ plugin for _esbuild_ bundler. There are some caveats with how the _esbuild_ plugin API works, and it seems that the consumer will have to provide a full path to the build directory directly to the plugin, because there's no other way to access the `build` path in the `onEnd` callback. I am using the `onEnd` callback in this case (which runs at the end of the build, implied by the name), as we first have to transform each `ts[x]/js[x]` before we can process the `StyleX` rules. Let me know if you have simpler ideas on how to solve the bundling of CSS.

An example of how consuming the plugin would look like:
```
await esbuild.build({
  entryPoints: [path.resolve(__dirname, 'src/index.js')],
  outdir: 'dist',
  outfile: 'bundled.js',
  plugins: [stylexPlugin({ absoluteFilePath: path.resolve(__dirname, 'dist/stylex.css') })],
});
```

Some remaining things to do:
- [x] Improve tests, specifically test for injections in dev and for the output in non-dev build
- [x] Add CSS writes to disk

nedjulius commented on packages/esbuild-plugin/src/index.js:
```diff
@@ -0,0 +1,105 @@
+
+        return {
+          contents: code,
+          loader: currFileName.endsWith('.js') ? 'js' : 'tsx',
```
`tsx` loader can handle `ts`, `tsx` and `jsx` files

nedjulius replied:
update: turns out this wasn't entirely correct. encountered some import resolution problems when using `tsx` loader for `JSX` files, even though the `esbuild` documentation said that either loader is fine. for the sake of simplicity, I just match extension to loader.

nmn commented on packages/esbuild-plugin/__tests__/__fixtures__/.babelrc.json:
```diff
@@ -0,0 +1,12 @@
+      "@babel/preset-env",
+      {
+        "targets": {
+          "ie": 11
```
Please use the same config as the other plugins.

nedjulius replied:
removed this fixture `babelrc` altogether, since its not used

nmn commented on packages/esbuild-plugin/__tests__/__fixtures__/index.js:
```diff
@@ -0,0 +1,39 @@
+});
+
+export default function App() {
+  return stylex(fooStyles.foo, styles.bar);
```
Use `stylex.props` instead.

nmn commented on packages/esbuild-plugin/__tests__/__fixtures__/index.js:
```diff
@@ -0,0 +1,39 @@
+    display: 'flex',
+    marginLeft: 10,
+    height: 700,
+    backgroundColor: 'red',
+    ':hover': {
+      backgroundColor: 'pink',
+    },
```
This is a legacy pattern. Use the officially supported API. 

```suggestion
    backgroundColor: {
      default: 'red',
      ':hover': 'pink',
    },
```

nmn commented on packages/esbuild-plugin/src/index.js:
```diff
@@ -0,0 +1,105 @@
+function stylexPlugin({
+  dev = IS_DEV_ENV,
+  unstable_moduleResolution = { type: 'commonJS', rootDir: process.cwd() },
+  stylexImports = ['stylex', '@stylexjs/stylex'],
```
```suggestion
  stylexImports = ['@stylexjs/stylex'],
```

nmn commented:
Looks promising so far!

nmn commented on packages/esbuild-plugin/src/index.js:
```diff
@@ -0,0 +1,110 @@
+            ...plugins,
+            /\.jsx?/.test(path.extname(currFileName))
+              ? flowSyntaxPlugin
+              : typescriptSyntaxPlugin,
+            jsxSyntaxPlugin,
```
I've run into probably because of this sometimes. So let's add it.

```suggestion
              : [typescriptSyntaxPlugin, {isTSX: true}],
            jsxSyntaxPlugin,
```

nedjulius commented on apps/esbuild-example/package.json:
```diff
@@ -0,0 +1,19 @@
+{
```
added an example package that provides a simple build script example for a React application. it is also easier to develop the plugin this way.

nmn commented:
@nedjulius Do you consider this PR ready for review?

nedjulius commented:
@nmn yeah, I think we can give it a go!

nmn commented on apps/esbuild-example/package.json:
```diff
@@ -0,0 +1,19 @@
+    "react-dom": "^18.2.0"
+  },
+  "devDependencies": {
+    "@stylexjs/esbuild-plugin": "^0.0.0",
```
Can you enable the ESLing plugin for this example as well? Like the NextJS and Docs apps.

nmn commented on apps/esbuild-example/public/index.html:
```diff
@@ -0,0 +1,19 @@
+    <title>@stylexjs/esbuild-plugin</title>
+    <meta charset="utf-8" />
+    <style>
+      body {
+        box-sizing: border-box;
+        padding: 0;
+        margin: 0;
+      }
```
Since you have `useCSSLayers: true`, the reset needs to be in a layer as well.

```suggestion
      @layer reset {
        html {
          box-sizing: border-box;
          padding: 0;
          margin: 0;
        }
      }
```

nmn commented on apps/esbuild-example/scripts/build.js:
```diff
@@ -0,0 +1,39 @@
+        absoluteFilePath: STYLEX_BUNDLE_PATH,
+        unstable_moduleResolution: {
+          type: 'commonJS',
+          rootDir: path.resolve(__dirname, '..'),
```
Change this to the stylex repo root dir?

nmn commented on apps/esbuild-example/src/App.jsx:
```diff
@@ -0,0 +1,47 @@
+import * as React from 'react';
+import ReactDOM from 'react-dom';
+import * as stylex from '@stylexjs/stylex';
+import { spacing, color } from './tokens.stylex.js';
```
...And then you can try using `open-props` here as well.

nmn commented on packages/esbuild-plugin/README.md:
```diff
@@ -0,0 +1,3 @@
+# @stylexjs/esbuild-plugin
+
+...
```
You can probably add some basic documentation now.

nmn commented on packages/esbuild-plugin/__tests__/index-test.js:
```diff
@@ -0,0 +1,202 @@
+  it('extracts and bundles CSS without inject calls, bundles JS', async () => {
+    const { js, css } = await build();
+
+    expect(js.text).toMatchInlineSnapshot(`
```
Can we use file snapshots here? They're pretty long.

nmn commented on packages/esbuild-plugin/src/index.js:
```diff
@@ -0,0 +1,118 @@
+/**
```
We're starting to add types to the packager plugins. See if you're up to the task of adding Flow types. If not, I'll do it after the fact.

nmn commented on packages/esbuild-plugin/src/index.js:
```diff
@@ -0,0 +1,118 @@
+  unstable_moduleResolution = { type: 'commonJS', rootDir: process.cwd() },
+  stylexImports = ['@stylexjs/stylex'],
+  // absolute path to bundled CSS file
+  absoluteFilePath = path.resolve(__dirname, 'stylex.css'),
```
Can we rename this option to something more descriptive?

nedjulius replied:
changed it to `generatedCSSFileName`. I don't think it should have the `absolute` in it's name as it can also consume relative file paths, though using absolute path is a recommendation.

nmn commented on packages/esbuild-plugin/src/index.js:
```diff
@@ -0,0 +1,118 @@
+  // absolute path to bundled CSS file
+  absoluteFilePath = path.resolve(__dirname, 'stylex.css'),
+  babelConfig: { plugins = [], presets = [] } = {},
+  useCSSLayers = false,
```
Again, let's leave this without a default value.

nmn commented on packages/esbuild-plugin/src/index.js:
```diff
@@ -0,0 +1,118 @@
+        });
+      });
+
+      onLoad({ filter: /\.[jt]sx?$/ }, async (args) => {
```
```suggestion
      onLoad({ filter: /\.[mc]?[jt]sx?$/ }, async (args) => {
```

To account for `.mjs` and `cjs` etc.

nmn commented on packages/esbuild-plugin/src/index.js:
```diff
@@ -0,0 +1,118 @@
+        const inputCode = await fs.readFile(currFilePath, 'utf8');
+
+        if (
+          !stylexImports.some((importName) => inputCode.includes(importName))
```
Oh I see why you have a default value for `stylexImports` now.

nmn commented on packages/esbuild-plugin/src/index.js:
```diff
@@ -0,0 +1,118 @@
+          plugins: [
+            ...plugins,
+            /\.jsx?/.test(currFileExtension)
+              ? flowSyntaxPlugin
```
In the non-TS, i.e. Flow case, we need to also use the `hermes-parser` plugin. This is something that needs to be added to each of the bundler plugins.

nmn commented on packages/esbuild-plugin/src/index.js:
```diff
@@ -0,0 +1,118 @@
+          return;
+        }
+
+        outputFiles.push({
+          path: '<stdout>',
+          contents: new TextEncoder().encode(collectedCSS),
+          get text() {
+            return collectedCSS;
+          },
+        });
```
What is going on here? Are you just printing the result to `stdout`?

nedjulius replied:
yeah. if the consumer specifies `write: false` in the `esbuild` configuration, then the bundled files are written to `outputFiles` array that contains objects of the structure above (they also have a `hash` property, which is generated by `esbuild` internally). just wanted to replicate the behavior for consistency, like if some plugin consumer decided to save `esbuild` output in memory first and then write everything to disk manually.

nmn reviewed (commented):
I don't see anything particularly wrong. Left a bunch nit-picky comments.

Will pull these changes locally to test how this works.

nedjulius commented:
@nmn I addressed all of your comments. the only issue left now is that `flow` complains about not being able to resolve some modules. not sure why though, but I assume it has something to do with those modules not having `flow` types (?). I saw that you created `flow_modules` file with type definitions for `rollup` and some `babel` plugins in the `rollup-plugin` -- is there a way to generate that somehow?

nmn commented:
> is there a way to generate that somehow?

There are some scripts online for it as well, but I manually translate the Typescript types for them.

nedjulius commented:
@nmn got you, thanks!

I added translations for _esbuild_ and also ported babel types from the _rollup-plugin_. perhaps `@babel/core` definition should be placed in `flow-typed` as it is now used in two multiple packages?

besides that, I also added the `node-libraries` to `flow-typed` as I was getting type resolution issues with filesystem module.

nedjulius commented on apps/docs/docs/learn/03-installation.mdx:
```diff
@@ -196,6 +196,48 @@ provides plugins for Webpack, Rollup, and Next.js.
 
 </details>
 
+<details>
```
Added _esbuild_ plugin example to the installations page in docs.

nmn replied:
Please be patient as it may take me a few days to review this properly.

nmn commented:
Lots of failing tests, so please work on that as I make my way through the PR.

nedjulius commented:
@nmn yup, looking into it. probably related to package lock merge conflicts that I tried to resolve yesterday.

nmn commented on flow-typed/node-libraries.js:
```diff
@@ -0,0 +1,15 @@
+/* eslint-disable ft-flow/no-types-missing-file-annotation */
+/* eslint-disable no-unused-vars */
+
+declare module 'fs/promises' {
```
Don't add this. The types for `fs` should already exist.

nmn commented on packages/esbuild-plugin/src/index.js:
```diff
@@ -0,0 +1,176 @@
+  stylexImports = ['@stylexjs/stylex'],
+  generatedCSSFileName = path.resolve(__dirname, 'stylex.css'),
+  babelConfig: { plugins = [], presets = [] } = {},
+  useCSSLayers,
```
Default this to `false`

nmn commented on packages/esbuild-plugin/src/index.js:
```diff
@@ -0,0 +1,176 @@
+  };
+}
+
+function getEsbuildLoader(fileName: string) {
```
Also, `.mjs` and `.cjs` and `.mts` and `.cts`

nedjulius replied:
`ts` loader handles `.mts` and `.cts` files, while `js` loader handles `.cjs` and `.mjs` -- [see here](https://esbuild.github.io/content-types/) for the reference

nmn commented on packages/esbuild-plugin/src/index.js:
```diff
@@ -0,0 +1,176 @@
+  return 'js';
+}
+
+function getFlowOrTypeScriptBabelSyntaxPlugins(fileName: string) {
```
We also need to add check to verify that we only attempt to transform JS files. This means ignoring `.json` files.

This regex might help.
```js
/[mc]?[jt]sx?$/
```

nedjulius replied:
the context in which this function is used guarantees that we only transform JS files:

https://github.com/facebook/stylex/blob/cd1f8b94a1856bad591d34dcfdd4e325416fa016/packages/esbuild-plugin/src/index.js#L93

I assume we don't want to allow processing invalid filetypes like `.mjsx` or `.cjsx`, hence I added an explicit filter above:

https://github.com/facebook/stylex/blob/cd1f8b94a1856bad591d34dcfdd4e325416fa016/packages/esbuild-plugin/src/index.js#L28

nmn reviewed (commented):
It's almost there! I've left a few comments. Please address those and I'll test it merge it within a few days.

nedjulius commented:
hey @nmn are you still considering adding this?

nmn commented:
@nedjulius Yes. Still planning to merge this for the next release.
